runtime/policies.go

```diff
@@ -0,0 +1,124 @@
+package runtime
+
+import (
+	"fmt"
+	"hash/maphash"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime/compilers/rillv1"
+	"go.uber.org/zap"
+)
+
+type policyEngine struct {
+	cache  *simplelru.LRU
+	lock   sync.Mutex
+	logger *zap.Logger
+}
+
+func newPolicyEngine(cacheSize int, logger *zap.Logger) *policyEngine {
+	cache, err := simplelru.NewLRU(cacheSize, nil)
+	if err != nil {
+		panic(err)
+	}
+	return &policyEngine{cache: cache, logger: logger}
+}
+
+var openPolicy = &ResolvedMetricsViewPolicy{
+	HasAccess: true,
+	Filter:    "",
+	Include:   nil,
+	Exclude:   nil,
+}
+
+type ResolvedMetricsViewPolicy struct {
+	HasAccess bool
+	Filter    string
+	Include   []string
+	Exclude   []string
+}
+
+func (p *policyEngine) resolveMetricsViewPolicy(attributes map[string]any, instanceID string, mv *runtimev1.MetricsView, lastUpdatedOn time.Time) (*ResolvedMetricsViewPolicy, error) {
+	if mv.Policy == nil {
+		return nil, nil
+	}
+
+	// if attributes is empty that means auth is disabled and also no user context is available
+	// since we are controlling the attributes we can safely return the open policy
+	if len(attributes) == 0 {
+		return openPolicy, nil
+	}
+
+	key := fmt.Sprintf("%v:%v:%v", instanceID, mv.Name, lastUpdatedOn)
+	for k, v := range attributes {
+		key += fmt.Sprintf(":%v:%v", k, v)
+	}
+	var h maphash.Hash
+	_, err := h.WriteString(key)
+	if err != nil {
+		return nil, err
+	}
+	cacheKey := h.Sum64()
```

Comment:
1. I don't believe map iteration order is deterministic
2. The maphash implementation is not a secure hash (so someone could get another user's access policy)
3. Should be able to write the individual components of the key directly to the hash (removing need for building a string first)